torch/distributed/pipelining/schedules.py

```diff
@@ -2735,6 +2736,257 @@ def _calculate_single_rank_operations(self, rank) -> list[Optional[_Action]]:
+                counters[backward_key] = backward_mb + 1
+
+                # Handle FULL_BACKWARD counter updates
+                input_key = (backward_stage, _ComputationType.BACKWARD_INPUT)
+                weight_key = (backward_stage, _ComputationType.BACKWARD_WEIGHT)
+                counters[input_key] = counters.get(input_key, 0) + 1
+                counters[weight_key] = counters.get(weight_key, 0) + 1
+
+                # Add the overlapped action
+                actions.append(
+                    _Action(-1, _ComputationType.OVERLAP_F_B, None, sub_actions)
+                )
+            else:
+                # Regular single action
+                key = (stage_index, computation_type)
+                mb_index = counters.get(key, 0)
+                actions.append(_Action(stage_index, computation_type, mb_index))
+                counters[key] = mb_index + 1
+
+                # If FULL_BACKWARD is used, also increment the separate BACKWARD_INPUT and BACKWARD_WEIGHT counters
+                # so they skip this microbatch since it's already covered by FULL_BACKWARD
+                if computation_type == _ComputationType.FULL_BACKWARD:
+                    input_key = (stage_index, _ComputationType.BACKWARD_INPUT)
+                    weight_key = (stage_index, _ComputationType.BACKWARD_WEIGHT)
+                    counters[input_key] = counters.get(input_key, 0) + 1
+                    counters[weight_key] = counters.get(weight_key, 0) + 1
+
+                # If BACKWARD_INPUT is updated, add corresponding weight action to queue and increment FULL_BACKWARD counter
+                if computation_type == _ComputationType.BACKWARD_INPUT:
+                    # Add weight action to queue for later processing
+                    weight_queue.append((stage_index, mb_index))
+                    full_backward_key = (stage_index, _ComputationType.FULL_BACKWARD)
+                    counters[full_backward_key] = counters.get(full_backward_key, 0) + 1
+
+        def add_weight_action(actions: list, counters: dict):
+            """Helper method to add a weight action from the queue."""
+            if not weight_queue:
+                return  # No pending weight actions, skip
+            # Pop the oldest weight action from the queue
+            actual_stage_index, weight_mb_index = weight_queue.pop(0)
+            actions.append(
+                _Action(
+                    actual_stage_index,
+                    _ComputationType.BACKWARD_WEIGHT,
+                    weight_mb_index,
+                )
+            )
+            # Update the counter for the actual stage that was processed
+            weight_key = (actual_stage_index, _ComputationType.BACKWARD_WEIGHT)
+            counters[weight_key] = counters.get(weight_key, 0) + 1
+
+        # Step 1: F0
+        step_1 = (num_ranks - rank - 1) * 2
+        for _ in range(step_1):
+            add_action(actions, stage0_index, _ComputationType.FORWARD, counters)
```

Comment:
nit: you could make this cleaner by not passing actions or counters into the helper. it can just access them anyway.